GLSP-102: Move tools and their feedback into dedicated modules

martin-fleck-at opened the pull request:
- Restructure code to separate tools module into dedicated modules
- Add custom methods to mouse and key tool to automate de-registration
- Return disposable from feedback dispatcher to return de-registration
- Extract parent class for creation tools
- Use disposable collections for cleanup in parent classes where useful

Fixes https://github.com/eclipse-glsp/glsp/issues/102

planger commented on packages/client/src/features/tools/change-bounds/change-bounds-tool.ts:
```diff
@@ -183,7 +167,7 @@ export class ChangeBoundsListener extends DragAwareMouseListener implements ISel
                 actions.push(...resizeActions);
             }
-            this.tool.dispatchFeedback(actions, this);
+            return actions;
```
Doesn't this undo the [earlier fix](https://github.com/eclipse-glsp/glsp-client/pull/262) showing moves as feedback (and thus re-apply on `UpdateModelAction`?

martin-fleck-at replied:
Maybe I merged something incorrectly, will double check! I tested some aspects but maybe not all.

martin-fleck-at replied:
Yeah, it was a bad merge, I'll restore the old behavior.

tortmayr commented on packages/client/src/container-modules.ts:
```diff
@@ -42,12 +42,19 @@ import modelHintsModule from './features/hints/di.config';
+import deletionToolModule from './features/tools/deletion/di.config';
+import edgeCreationToolModule from './features/tools/edge-creation/di.config';
+import edgeEditToolModule from './features/tools/edge-edit/di.config';
+import { enableDefaultToolsOnFocusLossModule } from './features/tools/enable-default-tools-on-focus-loss';
```
Unrelated, but maybe you could fix it while you are at it:
Looks like the `enable-default-tools-on-focus-loss` module is not default exported (in contrast to all other modules)

martin-fleck-at replied:
Ah yes, I'm always happy to provide more consistency!

planger commented on packages/client/src/features/tools/edge-creation/edge-creation-tool-feedback.ts:
```diff
@@ -0,0 +1,79 @@
+                return [MoveAction.create([{ elementId: edgeEnd.id, toPosition: anchor }], { animate: false })];
+            }
+        } else {
+            return [MoveAction.create([{ elementId: edgeEnd.id, toPosition: position }], { animate: false })];
```
Shouldn't this be registered as a feedback instead of directly returned?

planger commented on packages/client/src/features/tools/edge-creation/edge-creation-tool-feedback.ts:
```diff
@@ -0,0 +1,79 @@
+        if (endAtMousePosition instanceof SConnectableElement && edge.source && isBoundsAware(edge.source)) {
+            const anchor = this.computeAbsoluteAnchor(endAtMousePosition, Bounds.center(toAbsoluteBounds(edge.source)));
+            if (Point.euclideanDistance(anchor, edgeEnd.position) > 1) {
+                return [MoveAction.create([{ elementId: edgeEnd.id, toPosition: anchor }], { animate: false })];
```
Shouldn't this be registered as a feedback instead of directly returned?

tortmayr commented on packages/client/src/features/edit-label/edit-label-tool.ts:
```diff
@@ -37,14 +34,7 @@ export class DirectLabelEditTool extends BaseGLSPTool {
-
-    disable(): void {
-        this.keyTool.deregister(this.editLabelKeyListener);
-        this.mouseTool.deregister(this.editLabelMouseListener);
```
Clean refactor of the BaseGLSPTools. Love it!

tortmayr commented on packages/client/src/features/routing/di.config.ts:
```diff
@@ -0,0 +1,30 @@
+import { GLSPManhattanEdgeRouter } from './glsp-manhattan-edge-router';
+
+/**
+ * Registers the default tools of GLSP (node and edge creation, changing bounds, edge editing, deletion)
```
This comment looks like a copy&paste error

planger reviewed (commented):
I like this a lot, thank you very much!
I just had a few questions above inline, once I understand those, I'll do a bit more testing.
Thanks again!

tortmayr commented on packages/client/src/features/tool-feedback/di.config.ts:
```diff
@@ -14,48 +14,13 @@
-import { FeedbackEdgeEndView, SResizeHandleView } from './view';
+import { LocationPostprocessor, MoveCommand, TYPES, bindAsService, configureCommand } from '~glsp-sprotty';
 
 const toolFeedbackModule = new ContainerModule((bind, _unbind, isBound) => {
```
Shouldn't t his be part of the `base/feedback`? Since other modules also require the functionality it should not be part of an optional feature module.

martin-fleck-at replied:
I moved this around many times so I'm glad to take any good suggestion and that is definitely one! I'll move it.

tortmayr commented on packages/client/src/features/tools/base-glsp-tool.ts:
```diff
@@ -40,11 +47,11 @@ export abstract class BaseGLSPTool implements GLSPTool {
     }
 
-    deregisterFeedback(actions: Action[], feedbackEmitter?: IFeedbackEmitter): void {
+    deregisterFeedback(actions?: Action[], feedbackEmitter?: IFeedbackEmitter): void {
```
Parameter order should be reversed. Otherwise you have to call the method with undefined to deregister an emitter without actions. (`deregiserFeedback(undefined, someEmitter)`)
While we are at it we should also rename the `actions` parameter to `cleanupActions`

martin-fleck-at replied:
Very nice! You are right, I'll fix it!

tortmayr commented on packages/client/src/features/tools/change-bounds/change-bounds-tool-feedback.ts:
```diff
@@ -280,10 +280,12 @@ export class FeedbackMoveMouseListener extends MouseListener implements Disposab
             if (moveAction) {
                 result.push(moveAction);
             }
+            const feebackCleanup: Action[] = [];
```
```suggestion
            const feedbackCleanup: Action[] = [];
```
Although getting a fee back also sound good 😉

martin-fleck-at replied:
Ha! Good catch!

tortmayr commented on packages/client/src/features/tools/change-bounds/change-bounds-tool.ts:
```diff
@@ -326,23 +310,9 @@ export class ChangeBoundsListener extends DragAwareMouseListener implements ISel
-                this.tool.deregisterFeedback([resetResizeAction], this);
-            }
-            this.tool.deregisterFeedback([HideChangeBoundsToolResizeFeedbackAction.create()], this);
+            this.tool.registerFeedback([HideChangeBoundsToolResizeFeedbackAction.create()], this);
```
Shouldn't this be a `deregisterFeedback`?

martin-fleck-at replied:
Yes!

tortmayr commented on packages/client/src/features/tools/change-bounds/change-bounds-tool.ts:
```diff
@@ -183,7 +167,7 @@ export class ChangeBoundsListener extends DragAwareMouseListener implements ISel
                 actions.push(...resizeActions);
             }
-            this.tool.dispatchFeedback(actions, this);
+            return actions;
```
This does revert the change of https://github.com/eclipse-glsp/glsp-client/pull/262.
The idea was to dispatch the bounds changes as feedback so that they don't get reset if you receive a server update while resizing.

tortmayr reviewed (changes requested):
Awesome change and clean refactoring 😍 
I only had a code review for now and have some remarks.
Will do a complete test later today

martin-fleck-at commented on packages/client/src/base/view/glsp-mouse-tool.ts:
```diff
@@ -0,0 +1,25 @@
+import { Disposable, MouseListener, MouseTool } from '~glsp-sprotty';
+
+@injectable()
+export class GLSPMouseTool extends MouseTool {
```
Understood, I'll combine them into a single tool.

martin-fleck-at commented on packages/client/src/base/feedback/feedback-action-dispatcher.ts:
```diff
@@ -35,21 +35,22 @@ export interface IFeedbackActionDispatcher {
      * @param feedbackEmitter the emitter sending out feedback actions.
-     * @param actions the actions to be sent out.
+     * @param feedbackActions the actions to be sent out.
+     * @returns A 'Disposable' that de-registers the `feedbackEmitter` with the given `cleanupActions`.
```
Good catch, thank you!

martin-fleck-at commented on packages/client/src/features/tools/base-glsp-tool.ts:
```diff
@@ -14,21 +14,28 @@
+    @inject(GLSPKeyTool) protected keyTool: GLSPKeyTool;
     @inject(EditorContextService) protected readonly editorContext: EditorContextService;
 
+    protected onDisable = new DisposableCollection();
```
`toDispose` is a bit too generic for my taste but how about `toDisposeOnDisable`? This is a similar naming convention that is used in Theia where they have `toDisposeOnDetach` on widgets.

martin-fleck-at commented on packages/client/src/base/feedback/feedback-action-dispatcher.ts:
```diff
@@ -59,41 +60,44 @@ export interface IFeedbackActionDispatcher {
         return result;
     }
+
+    protected dispatchFeedback(actions: Action[], feedbackEmitter: IFeedbackEmitter): void {
```
Good catch, I'll change it.

martin-fleck-at commented:
@planger @tortmayr Thank you very much for the fast feedback! I addressed all concerns. I do feel there is still something ab it off with the change bounds tool when you resize an element and then click somewhere else but that is also true on master as far as I can see. Maybe one of you has any pointers or it is a non-issue after all ;-)

martin-fleck-at commented:
@tortmayr Thank you! I added another commit that should fix all the behavior now.

tortmayr commented:
Thanks for the fast fix. 
Works almost perfectly now but I noticed on other issue:
Clicking on the tool palette after an resize (whilest the resize handles are still shown) results 
in unintended behavior. Seems like there is a feedback action dispatched that resizes the element to its previous state. On the next server update (e.g. by creating a node) the actual size is restored.
![Peek 2023-07-13 13-04](https://github.com/eclipse-glsp/glsp-client/assets/2311075/c0ab4c28-9d31-4ce2-b464-a6c7bd46cf19)

martin-fleck-at commented:
@tortmayr Good catch! With the latest fix everything should work as expected.
